pkg/appgw/backendaddresspools.go

```diff
@@ -7,82 +7,126 @@ package appgw
-		endpoints := builder.k8sContext.GetEndpointsByService(backendID.serviceKey())
-		if endpoints == nil {
-			logLine := fmt.Sprintf("Unable to get endpoints for service key [%s]", backendID.serviceKey())
-			builder.recorder.Event(backendID.Ingress, v1.EventTypeWarning, "EndpointsEmpty", logLine)
-			glog.Warning(logLine)
-			builder.backendPoolMap[backendID] = &emptyPool
-			continue
+func (builder *appGwConfigBuilder) BackendAddressPools(ingressList []*v1beta1.Ingress) (ConfigBuilder, error) {
+	defaultPool := defaultBackendAddressPool()
+	addressPools := map[string]*n.ApplicationGatewayBackendAddressPool{
+		*defaultPool.Name: defaultPool,
+	}
+	for backendID, serviceBackendPair := range builder.getServiceBackendPairMap() {
+		if pool := builder.getBackendAddressPool(backendID, serviceBackendPair, addressPools); pool != nil {
+			// TODO(draychev): deprecate the caching of state in builder.backendPoolMap
+			builder.backendPoolMap[backendID] = pool
+			addressPools[*pool.Name] = pool
 		}
+	}
+	builder.appGwConfig.BackendAddressPools = getBackendPoolMapValues(&addressPools)
+	return builder, nil
+}
 
-		for _, subset := range endpoints.Subsets {
-			endpointsPortsSet := make(map[int32]interface{})
-			for _, endpointsPort := range subset.Ports {
-				if endpointsPort.Protocol != v1.ProtocolTCP {
-					continue
-				}
-				endpointsPortsSet[endpointsPort.Port] = nil
-			}
+func getBackendPoolMapValues(m *map[string]*n.ApplicationGatewayBackendAddressPool) *[]n.ApplicationGatewayBackendAddressPool {
+	var backendAddressPools []n.ApplicationGatewayBackendAddressPool
+	for _, addr := range *m {
+		backendAddressPools = append(backendAddressPools, *addr)
+	}
+	return &backendAddressPools
+}
 
-			if _, ok := endpointsPortsSet[serviceBackendPair.BackendPort]; ok {
-				addressPoolName := generateAddressPoolName(backendID.serviceFullName(), backendID.Backend.ServicePort.String(), serviceBackendPair.BackendPort)
-				// The same service might be referenced in multiple ingress resources, this might result in multiple `serviceBackendPairMap` having the same service key but different
-				// ingress resource. Thus, while generating the backend address pool, we should make sure that we are generating unique backend address pools.
-				addressPool, ok := addressPools[addressPoolName]
+func (builder *appGwConfigBuilder) getBackendAddressPool(backendID backendIdentifier, serviceBackendPair serviceBackendPortPair, addressPools map[string]*n.ApplicationGatewayBackendAddressPool) *n.ApplicationGatewayBackendAddressPool {
+	endpoints := builder.k8sContext.GetEndpointsByService(backendID.serviceKey())
+	if endpoints == nil {
+		logLine := fmt.Sprintf("Unable to get endpoints for service key [%s]", backendID.serviceKey())
+		// TODO(draychev): Move "reason" into an enum
+		builder.recorder.Event(backendID.Ingress, v1.EventTypeWarning, "EndpointsEmpty", logLine)
+		glog.Warning(logLine)
+		return defaultBackendAddressPool()
```

Comment:
We can combine `return nil` with `builder.backendPoolMap[backendID] = defaultPool`?

Reply:
I like this!